### PR TITLE
Fix ASAN identified global-buffer-overflow

### LINKED ISSUE #690: `sh_setenviron()` can be called with a NULL pointer
The old libast implementation of `setenviron()` and the ksh `sh_setenviron()` variant support being called with a NULL pointer. But the ksh version doesn't do anything when passed NULL. There does not appear to be any good reason to still support calling those functions with NULL. The only place which does so is the `procopen()` function. Which is used by the coshell subsystem which is no longer enabled by default. it is also used by the standalone `pty` command (which I'd like to remove, see #634). The former by definition doesn't need this feature since it uses the ksh implementation which treats NULL as a no-op. And it is not obvious the `pty` command needs that optimization.

### PR DESCRIPTION
**Pull request by krader1961**
Fixes #690

### CODE REVIEW

**Comment by krader1961** [comment]
Note that this merely eliminates one bug detected by ASAN in this module. It allows ASAN to identify a different bug (since ASAN kills the process when the first bug is detected). I added a big fat comment because my transformation avoids the invalid indexing past the end of a global string and should be functionally equivalent to the original code. But I have absolutely no idea what the block of code in question is intended to do. Assuming the intent of the original code is correct and was wrong simply because the author was being too clever with a manual optimization this change is correct and won't introduce or mask other bugs.

**Comment by krader1961** [comment]
I wanted to see if the block of code predicated on the invalid array index was ever executed. So I added this to the inner block of the original and revised version of the code:

```
fprintf(stderr, "WTF size %d  up->cp |%s|\n", numeric, up->cp);
```

The results are identical and show that this is only ever exercised when called from the signal handling path and the strings in question are "SI_USER", "EXITED", "STOPPED" and the like. Furthermore, if I simply comment out this block of code (both original and new) not a single unit test is affected.

This block of code does exist in the 93u+ branch. Whether or not it did something useful in that branch is unknown but it appears to be broken there as well by reading past the end of a string. Because it exists in that branch I am hesitant to remove it even though it appears to be pointless as well as wrong in its original form.

**Comment by krader1961** [comment]
Gah! I'll fix the commit comment. It was supposed to say `Fixes #691`.

File: src/cmd/ksh93/sh/name.c [hunk 1]
```diff
@@ -2424,13 +2424,21 @@ char *nv_getval(Namval_t *np) {
...
-        return cp;
+
+    // See issue #690. The original version of this block of code did an invalid `up->cp[numeric]`.
+    // That is invalid because it assumes the string pointed to by `up->cp` is at least `numeric`
+    // (i.e., `nv_size(np)`) chars long. That is an invalid assumption. This version avoids a
+    // possible incorrect reference past the end of the string but it's not obvious what this block
+    // is meant to do and may therefore still be incorrect.
+    if (up->cp && !nv_isattr(np, NV_LJUST | NV_RJUST)) {
+        int size = nv_size(np);
+        if (size > 0 && size != strlen(up->cp)) {
```
[the remark this answers is not in the document]
Comment (krader1961) [reply src/cmd/ksh93/sh/name.c]: That's a good question. The original code seems to be checking if the the string pointed to by `up->cp` is longer than the length of the name/value pair. It did so via the incorrect `&& up->cp[numeric]` expression. That expression assumes the string will be either the length of the namval object, in which case `up->cp[numeric]` refers to the terminating zero byte of the string, and the expression is false, or the string is longer than that and the expression is true. The problem is that every single time this code is executed via a unit test the string is shorter than `numeric`. In which case `up->cp[numeric]` refers to a byte past the end of the string and the result is unspecified. It's a [HCF](https://en.wikipedia.org/wiki/Halt_and_Catch_Fire) instruction.

The new expression should really be `size > strlen(up->cp)` to match the intent of the original version. I'll fix that. but having said that I still don't understand why this block of code exists. It is never exercised except by accident due to the invalid HCF hand optimization in the original code.

**Comment by siteshwar** [comment]
LGTM

**Pull request closed without merging**